data/tutorials/wf_03_metaprogramming.md

```diff
@@ -0,0 +1,452 @@
+would look for the type in the file, and serialize it differently depending on
+the structure of the type, that is whether it is a variants type, a record type,
+the structure of its subtypes...
+
+But, to understand the structure of the type, the preprocessor need to parse
+those information. Given that parsing OCaml syntax is complex, it is not
+something that we want to do do again and again whenever we write a new
+preprocessor.
+
+The solution to this is to use PPXes, preprocessors that run on the result of
+the parsing of the file.
+
+## PPXes
+
+PPxes are a different kind of preprocessors, that do not run on the textual
+source code, but on the result of the parsing: the Abstract Syntax Tree (AST),
+or more precisely in OCaml, the parsetree. In order to understand PPXes well, we
+need to understand what is this parsetree.
+
+### OCaml's parsetree, the first OCaml's AST
+
+During the compilation phase, OCaml's compiler will parse the input file into an
+internal representation of it, called the parsetree. The program is represented
+as a tree, with a very complicated OCaml type, that you can find in the
+[`Parsetree` module](https://v2.ocaml.org/api/compilerlibref/Parsetree.html).
+
+Let us give already a few properties of this tree:
+- Each node in the AST has a type corresponding to a different role, such as
+  "let definitions", "expressions", "patterns", ...
+- The root of the tree is a list of `structure_item`s
+- A `structure_item` can either represent a toplevel expression, a type
+  definition, a let definition, ... This is determined using a variant type.
+
+There are two complementary ways of getting a grasp on the parsetree type. One
+is to read the API documentation, which include examples of what each type and
+value represent. The other is to make OCaml dump the parsetree value of crafted
+OCaml code. This is possible thanks to the option `-dparsetree`, available in
+the `ocaml` toplevel (as well as in `utop`):
+
+```shell
+
+$ ocaml -dparsetree
+[Omitted output]
+# let x = 1 + 2 ;;
+
+Ptop_def
+  [
+    structure_item (//toplevel//[1,0+0]..[1,0+13])
+      Pstr_value Nonrec
+      [
+        <def>
+          pattern (//toplevel//[1,0+4]..[1,0+5])
+            Ppat_var "x" (//toplevel//[1,0+4]..[1,0+5])
+          expression (//toplevel//[1,0+8]..[1,0+13])
+            Pexp_apply
+            expression (//toplevel//[1,0+10]..[1,0+11])
+              Pexp_ident "+" (//toplevel//[1,0+10]..[1,0+11])
+            [
+              <arg>
+              Nolabel
+                expression (//toplevel//[1,0+8]..[1,0+9])
+                  Pexp_constant PConst_int (1,None)
+              <arg>
+              Nolabel
+                expression (//toplevel//[1,0+12]..[1,0+13])
+                  Pexp_constant PConst_int (2,None)
+            ]
+      ]
+  ]
+
+val x : int = 3
+```
+
+Note that the parsetree is an internal representation of the code that happens
```

Comment:
```suggestion
Note that the Parsetree is an internal representation of the code that happens
```
Other OCaml documentation shows this capitalized, but please let me know if that's incorrect (or correct in some instances and other others) and I'll add that to the style guide.

Reply:
I have no idea what it should be. I know it has to be capitalized for the `Parsetree` OCaml module, so maybe it is better to capitalize it everywhere...